### ISSUE TITLE
Missing constraints

### ISSUE THREAD

**Issue by pgrinaway** [choderalab/perses#142]
I seem to occasionally have a problem with missing constraints. That is, `GeometryEngine` looks for bond parameters, and seeing none, goes off and looks for a bond constraint in the unmodified system. However, there are often times when there is no constraint between the atoms of interest, leading to an exception. Reproduced with the code below.
```python
def test_logp_reverse():
    np.seterr(all='raise')
    molecule_name_1 = 'propane'
    molecule_name_2 = 'ethane'

    molecule1 = generate_initial_molecule(molecule_name_1)
    molecule2 = generate_initial_molecule(molecule_name_2)
    new_to_old_atom_mapping = align_molecules(molecule1, molecule2)

    sys1, pos1, top1 = oemol_to_openmm_system(molecule1, molecule_name_1)
    sys2, pos2, top2 = oemol_to_openmm_system(molecule2, molecule_name_2)

    import perses.rjmc.geometry as geometry
    import perses.rjmc.topology_proposal as topology_proposal

    sm_top_proposal = topology_proposal.TopologyProposal(new_topology=top2, new_system=sys2, old_topology=top1, old_system=sys1,
                                                                    logp_proposal=0.0, new_to_old_atom_map=new_to_old_atom_mapping, new_chemical_state_key="CCC", old_chemical_state_key="CC", metadata={'test':0.0})
    geometry_engine = geometry.FFAllAngleGeometryEngine({'test': 'true'})
    new_positions, logp_proposal = geometry_engine.propose(sm_top_proposal, pos1, beta)

    logp_reverse = geometry_engine.logp_reverse(sm_top_proposal, new_positions, pos1, beta)
    print(logp_proposal)
    print(logp_reverse)
    print(logp_reverse-logp_proposal)
```

I don't think this is `GeometryEngine`'s fault, since it's the system that it accepts as input, and inspection reveals a lack of constraints.

**Comment by jchodera** [comment]
I think I've turned off constraints in the `SystemGenerator` for everything in `testsystems` because of some issues we were having before.  Should this cause a problem?  Or is it that there is some inconsistency between the `SystemGenerator` (which may not be using constraints) and the initial `System` object (which may have been accidentally constructed with constraints)?  Or potentially some missing bonds in `Topology`?

**Comment by pgrinaway** [comment]
Hm, well, my test function does have constraints, but even if they were off, then there should be a harmonic bond term between a methyl carbon and a bonded hydrogen, no?

**Comment by pgrinaway** [comment]
Let me retry without constraints.

**Comment by jchodera** [comment]
If constraints are off, there should definitely be `HarmonicBondForce` terms.

**Comment by pgrinaway** [comment]
Ok, without the constraints, there are no issues. But I don't think this is a problem of `GeometryEngine` not perceiving constraints--some of them simply aren't there (and there is no `HarmonicBondForce` term either).

**Comment by pgrinaway** [comment]
I'd say maybe I'm missing bonds in the topology, but parmed does perceive a bond, just no parameters.

**Comment by jchodera** [comment]
Could it be that somehow we're not generating the appropriate `HarmonicBondForce` terms in the `gaffTemplateGenerator`?

**Comment by jchodera** [comment]
I think the easiest way to debug would be to write a nosetest that also spits out, for a particular large-to-small molecule transformation:
* PDB file of the small molecule (from the `Topology` and `positions`)
* serialized XML file of `System` that is problematic

If you could also add the tripos atom names to the molecule first, that would be awesome.

**Comment by jchodera** [comment]
I tried adding some test code to `SystemGenerator.create_system()` that checks that the `System` has the same bonds as `Topology`:
```
        # Check to make sure the system has the same bonds as topology
        system_bonds = set()
        for index in range(system.getNumConstraints()):
            atom1_index, atom2_index, r0 = system.getConstraintParameters(index)
            system_bonds.add((atom1_index, atom2_index))
            system_bonds.add((atom2_index, atom1_index))
        for force_index in range(system.getNumForces()):
            force = system.getForce(force_index)
            if force.__class__.__name__ in ['HarmonicBondForce', 'CustomBondForce']:
                for index in range(force.getNumBonds()):
                    result = force.getBondParameters(index)
                    [atom1, atom2] = result[0], result[1]
                    system_bonds.add((atom1_index, atom2_index))
                    system_bonds.add((atom2_index, atom1_index))
        topology_bonds = set()
        for (atom1, atom2) in topology.bonds():
            topology_bonds.add((atom1.index, atom2.index))
            topology_bonds.add((atom2.index, atom1.index))
        if system_bonds != topology_bonds:
            msg  = 'System and Topology have different bonds.\n'
            msg += 'Bonds in System not in Topology: %s\n' % str(system_bonds.difference(topology_bonds))
            msg += 'Bonds in Topology not in System: %s\n' % str(topology_bonds.difference(system_bonds))
            msg += 'System written to BuildSystem-failure.pdb\n'
            # Write PDB file of failed topology
            from simtk import unit
            nparticles = sum([1 for atom in topology.atoms()])
            positions = unit.Quantity(np.zeros([nparticles,3], np.float32), unit.angstroms)
            from simtk.openmm.app import PDBFile
            outfile = open('BuildSystem-failure.pdb', 'w')
            pdbfile = PDBFile.writeFile(topology, positions, outfile)
            outfile.close()
            raise Exception(msg)
```
Unfortunately, this just picks up the fact that the `System` seems to have `H-H` bonds in water, while the `Topology` doesn't, so I don't think this is useful.

**Comment by pgrinaway** [comment]
Resolved in #252 , reopen if necessary
